File: cpp/233.Number_of_Digit_One.cpp
// https://leetcode.com/articles/number-of-digit-one/
class Solution {
public:
    int countDigitOne(int n) {
        int cnt = 0;
        for(long long i = 1; i <= n; i *= 10) {
            long long div = i * 10;
            cnt += (n / div) * i + min(i, max(0LL, n % div - i + 1));
        }
        return cnt;
    }
};

File: go/233.Number_of_Digit_One.go
// https://leetcode.com/articles/number-of-digit-one/
func countDigitOne(n int) int {
	cnt := 0
	for i := 1; i <= n; i *= 10 {
		divider := i * 10
		tmp := 0
		if (n%divider - i + 1) > 0 {
			tmp = (n%divider - i + 1)
		}
		if tmp > i {
			tmp = i
		}
		cnt += (n/divider)*i + tmp
	}
	return cnt
}
